lib/Parse/ParseType.cpp

```diff
@@ -196,7 +196,12 @@ ParserResult<TypeRepr> Parser::parseTypeSimple(Diag<> MessageID,
     }
     LLVM_FALLTHROUGH;
   default:
-    diagnose(Tok, MessageID);
+    auto diag = diagnose(Tok, MessageID);
+    // If the next token is closing or separating, the type was likely forgotten
+    if (Tok.isAny(tok::r_paren, tok::r_brace, tok::r_square, tok::arrow,
+                  tok::equal, tok::comma, tok::semi)) {
+      diag.fixItInsert(getEndOfPreviousLoc(), " <#type#>");
```

Comment:
This isn't perfect because the space isn't always needed, but I can't think of a way to test whether it's needed without looking at the *previous* token, which is a bit tricky. I guess this is good enough!